### DISCUSSION TITLE
allow permits keyword for sealed class declarations (and interfaces)

### DISCUSSION THREAD

**Discussion by CosminSontu** [dotnet/csharplang#7952] [Language Ideas]
Java allows defining sealed classes and interfaces.
With [JEP-409](https://openjdk.org/jeps/409) Java introduced the ability to specify exceptions (classes that can actually extend/implement the sealed classes/iterfaces) via permits keyword.

This feature is not that complex and offers quick gains in terms of pattern matching exhaustiveness.
Would this be a relatively easy to implement language feature that would provide quick wins while also avoiding unnecessarily complicating the language syntax?

I know it's a Java feature, but it's good. No worries, Java borrowed from C# lots of recent features so this is actually returning compliments.
PS: I have done a search for similar issues but in case I missed those, feel free to point them out.

**Comment by CosminSontu** [comment]
the conept could be extended to structs for uniformity.
Spotted this issue post-factum: https://github.com/dotnet/csharplang/issues/485. No mention of Java's approach so maybe this is useful...

**Comment by huoyaoyuan** [comment]
Note that the word "sealed" means completely different thing in C#.

**Reply by HaloFour** [reply]
It means the same thing in Java.  The difference is that with the `permits` clause you are saying that the type is sealed (cannot be extended), except for the specific exceptions.

It is a nice solution, and it avoids the problem of forcing the case classes to be nested.

**Reply by huoyaoyuan** [reply]
At the view of runtime, sealed class is special about casting and some other optimizations.

**Reply by HaloFour** [reply]
> At the view of runtime, sealed class is special about casting and some other optimizations.

Those would be implementation details and not relevant to a discussion specifically around a language feature.  Obviously the runtime would require changes in order to support a feature like this, and part of that would impact those implementation details.  I would expect that any existing `sealed` type without a `permits` clause would behave exactly as it does today as the runtime could still make the same assumptions as it does today.

**Reply by CosminSontu** [reply]
For a wider perspective, the art of possible with a feature like _permits_ : https://www.youtube.com/watch?v=8FRU_aGY4mY&t=714s

**Reply by KennethHoff** [reply]
IIRC, Java kind of has two keywords that does basically the same thing. `final` which is identical to `sealed` in C#, and `sealed` which is also identical to `sealed` in C# except that it can allow specific classes to extend them using the `permits` keyword.

I do not understand why there are two keywords that do the same exact thing in Java, but that's Java I guess :shrug:. Why not just this:

```java
// You're supposed to use `final` instead in this scenario, despite it - for all intents and purposes - does the same exact thing
public sealed class Side;

// Using `final` here however is invalid.
public sealed class Side permits Left, Right;
```

Again, my understanding of Java might be wrong.

**Reply by CosminSontu** [reply]
My understanding is that final in java can be used for : 
1) variables -> akin to readonly keyword in C#
2) methods -> makes method non-overridable, similar to sealed in c# but only to stop propagating overridability since non-static methods are sealed by default (and need to be marked as virtual to become overridable)
3) classes -> here c# uses sealed

Now sealed in java was introduced in the aforementioned JEP-409 to differentiate from final usage and maybe to prevent breaking changes... so sealed was introduced so that you can prevent inheriting but with some **_permitted_** exceptions.

Java seems to have the following classifications for class inheritance:
1) final -> extending is not permitted
2) sealed -> extending is permitted for explicit list of exceptions
3) non-sealed -> extending is permitted by unknown subclasses

Why this is necessary, escapes me... since they could have had:
1) final
2) final with permits
3) non final

It would be nice to be able to pair permits with sealed in c#.
1) sealed
2) sealed with permits
3) non sealed (default)

**Reply by CosminSontu** [reply]
A beautiful side-effect of a sealed/permits - based inheritance hierarchy is that
implementing a sealed class (because you are permitted to :) ) you now have the following options:
1) to disallow further specialization, by just using sealed modifier (without permits)
2) to allow further specialization but within specified permitted implementations (with permits)
3) to allow unbounded specialization by not specifying sealed at all.  ( here pattern matching would ask for a wildcard match arm ...)

**Reply by HaloFour** [reply]
This is something I really like about Java's approach, it allows further extending from the permitted cases to create a hierarchy but still retaining the exhaustiveness checking.

**Reply by KennethHoff** [reply]
Is this problem space also solved by discriminated unions? Personally I think this is a very clever solution to the problem of exhaustiveness, but does it add anything else that's not also covered by union types, which is actively on the working set and additionally allows for "closed type hierarchies" of types you do _not_ personally own.

How would this interact with extension interfaces? Could you create a sealed interface hierarchy containing extension interfaces? That's basically a type union right there 😜

**Reply by HaloFour** [reply]
There is a lot of overlap.  This is how Java chose to implement ADTs.  For "reference-based" DUs, a closed type hierarchy fits the bill nicely.  But closing that hierarchy without a feature like "permits" forces a convention on how that hierarchy would be defined.  The current proposal inclused nested types which are sealed, forbidding further extension, which Java does allow.

**Reply by CosminSontu** [reply]
The question is... Is this something that could land sooner than Roles and Extensions while not causing any interference... while allowing Roles and Extensions enough time and space to evolve? 
Sealed/permits seems to be compatible with roles at least...

**Comment by jos3s** [comment]
Any updates on this feature? Studying a few months ago I found a case that would like to have this possibility that java offers, but I did not find much information about its equivalent in C# until I found this discussion

**Reply by CyrusNajmabadi** [reply]
@jos3s we're looking at exhaustiveness in our DU efforts.

**Comment by nathan130200** [comment]
One good thing i found in JEP-409 is about extending StringBuilder. When i need reuse string builder (eg.: some kind of `StringBuilderPool`) i need implement a wrapper class that holds a StringBuilder reference via some property or `Get` method, because in C# string builder is a sealed type.
